packages/devtools_app/lib/src/screens/memory/panes/chart/controller/chart_connection.dart

```diff
@@ -36,30 +38,67 @@ class ChartConnection extends DisposableController
   );
 
   Timer? _pollingTimer;
-  bool _connected = false;
 
-  late final isDeviceAndroid =
-      serviceConnection.serviceManager.vm?.operatingSystem == 'android';
+  /// If completed, this instance was connected to the application.
+  final Completer<void> _connected = Completer();
+
+  /// If true, the instance lost connection to the application.
+  final Completer<void> _disconnected = Completer();
+
+  late final isDeviceAndroid = _isDevToolsCurrentlyConnected()
+      ? serviceConnection.serviceManager.vm?.operatingSystem == 'android'
+      : false;
+
+  /// True if DevTools is in connected mode and the connection to the app is alive.
+  bool _isDevToolsCurrentlyConnected() =>
+      // Theoretically it should be enough to check only connectedState.value.connected,
+      // but practically these values are not always in sync and, if at least
+      // on of them means disconnection this class consider the connection as lost,
+      // and stops interaction with it.
+      !offlineDataController.showingOfflineData.value &&
+      serviceConnection.serviceManager.connectedState.value.connected &&
+      serviceConnection.serviceManager.connectedApp != null;
 
   Future<void> maybeConnect() async {
-    if (_connected) return;
-    await serviceConnection.serviceManager.onServiceAvailable;
-    autoDisposeStreamSubscription(
-      serviceConnection.serviceManager.service!.onExtensionEvent
-          .listen(_memoryTracker.onMemoryData),
-    );
-    autoDisposeStreamSubscription(
-      serviceConnection.serviceManager.service!.onGCEvent
-          .listen(_memoryTracker.onGCEvent),
-    );
-    await _onPoll();
-    _connected = true;
+    if (_connected.isCompleted || _disconnected.isCompleted) return;
+    _connected.complete();
+    await _runSafely(() async {
+      await serviceConnection.serviceManager.onServiceAvailable;
+      autoDisposeStreamSubscription(
+        serviceConnection.serviceManager.service!.onExtensionEvent
+            .listen(_memoryTracker.onMemoryData),
+      );
+      autoDisposeStreamSubscription(
+        serviceConnection.serviceManager.service!.onGCEvent
+            .listen(_memoryTracker.onGCEvent),
+      );
+      await _onPoll();
+    });
   }
 
   Future<void> _onPoll() async {
-    _pollingTimer = null;
-    await _memoryTracker.pollMemory();
-    _pollingTimer = Timer(chartUpdateDelay, _onPoll);
+    assert(_connected.isCompleted);
+    if (_disconnected.isCompleted) return;
+    await _runSafely(() async {
+      _pollingTimer = null;
+      await _memoryTracker.pollMemory();
+      _pollingTimer = Timer(chartUpdateDelay, _onPoll);
```

Comment:
two things here:

1. Shouldn't this be `Timer.periodic` so we are actually polling and not just calling this twice?
2. I think the callback is supposed to be `_memoryTracker.pollMemory`, not `_onPoll`. Otherwise this just resets the value of the `_pollingTimer` var. Instead of using recursion, `Timer.periodic` is what we want here to implement polling.

Reply:
The timer was not periodic in the original code, and it was making sense as we do not want polls to have risk to overlap.

Reply:
Added comment.